### ISSUE TITLE
Migrate Smarty 2 code to Smarty 3

### ISSUE THREAD

**Issue by ghost** [openemr/openemr#2330]
We have this previous issue: https://github.com/openemr/openemr/issues/799

BACKGROUND: Moving from Smarty 2 to Smarty 3 seems like an easy win. A good amount of work for sure, but at least straightforward. Smarty 2 is from 2007, Smarty 3 is more like 2014ish.

TWIG: Someone told me to just dump Smarty and move all of it to Twig. This option is appealing as all work has stopped on Smarty with the exception of security updates here and there. We are also using Twig in other parts of the codebase. 

FINAL CONCLUSION: probably just migrate from Smarty 2 to Smarty 3. This will pay a lot of benefits. Maybe move to Twig later but save that discussion for another day.

Thoughts on final conclusion?
Any idea on how to get started on this work? Specific folders?

**Comment by bradymiller** [comment]
There are no use cases for twig in the codebase (the twig "wiring" is there, but not actually even used). Converted all smarty to twig would be a monumental feat (note the gacl advanced admin module is all in smarty). Going from smarty2 to smarty3 is way for manageable of a project, so definitely recommend going that route. To start, recommend identifying the places smarty is used (hint; there are 3 places) and then would go from there.

**Comment by danehrlich1** [comment]
Is this a 10, 20, or 40 hour project do you think? Thanks.

**Comment by bradymiller** [comment]
hi @danehrlich1 ,
It will depend how smoothly it goes (one of the places, the gacl admin module, may require no work except for updating the smarty version in composer, however the other 2 places, calendar and core, will likely be more bothersome). I would guess 10-30 hours depending on how it goes.

**Comment by danehrlich1** [comment]
Anthony Zullo said he'd do this, and his work is solid. I think will take him 2 months or so as he travels, but I think we can safely consider this done :)

[pops champagne]

**Comment by danehrlich1** [comment]
@anromeo @bradymiller Is this still worth working on? Even with all the calendar code you had to take out?

**Comment by bradymiller** [comment]
yes, still have calendar code, facilities code, and phpgacl code that uses this. The yanking of most of the calendar code should make it easier though :)

**Comment by danehrlich1** [comment]
Anthony:

Status? I will still keep contract the same even though there should be
less work for you now if you can finish by end of June. That's about 2
weeks.

**Comment by anromeo** [comment]
Hi Dan!

I probably need that week or two, but I'll finish before July. I just have
been so overcommitted recently.

Kind regards,
Anthony

**Comment by bradymiller** [comment]
hi @anromeo , I'm gonna open up this issue for new developers. Just making sure you are no longer working on this. thanks, -brady

**Comment by thesmallstar** [comment]
sir @bradymiller, I am planning to include this issue in my GSOC proposal, can I give it an early start?
Also since you estimated this to be a 30hrs (+20 hrs) job including only this won't make much sense for 3 months GSOC project, so I am planning to work on a new feature + this, along with solving normal-high priority issue. 

What do you suggest?

**Comment by bradymiller** [comment]
hi @thesmallstar , I am bad at estimating hours and that was more in terms with to just get it working. If done well, would likely be many more hours. And it could definitely be incorporated within a modernization GSOC proposal. Happy to assign it to you so you can look into it more and perhaps tackle it early.

**Comment by thesmallstar** [comment]
sir @bradymiller, can you give some initial guidance on how to get started and tell me if there is anything I should know? I will be good to go on my own then, I plan to complete the most of the part of this before GSOC(before may), so that if I am selected in GSOC I can invest more time in creating something/adding a new feature. 
thanks

**Comment by thesmallstar** [comment]
@bradymiller  I am thinking to start with this section(admin -> acl)
![image](https://user-images.githubusercontent.com/42006277/75607182-4863d180-5b1a-11ea-985f-65753ba9574a.png)
I have found the code that is to be changed and looks good to start the work. 
Do you have any suggestions?

**Comment by bradymiller** [comment]
hi @thesmallstart ,

There are 3 places that use Smarty in OpenEMR each with different challenges (search for '/documents/smarty' to find them).

1. php-gacl module - this will likely be the easiest one and may already work with Smarty3

2. core OpenEMR - this will be moderate difficulty likely. these templates are in the /templates directory. there has been work in the past to remove all php tags from these templates, so this will likely involve pluggin in Smarty 3, turning on debug and working through all the issues that come up.

3. calendar - this will be the toughest. Note there are numerous places in the calendar templates that use php tags to break the templating. this is important since smarty3 is not supposed to allow this (note there is a way to initiate smarty3 to be more flexible in this regard). So the steps could involve getting it to work in smarty3 with the workaround and allowing the php tags, and then reworking the code to remove the php tags (ie. places that are not doing standard templating).

Then something to think about if have time would be converting 2 (and maybe 3) to twig templating (this would be more in line with a full GSoC project). OpenEMR has the pieces to support twig in the codebase already, and there is an example of how we plan to use it in this PR:
https://github.com/openemr/openemr/pull/2862

**Comment by thesmallstar** [comment]
Thank you, I will go through this. Will start with the testing for the easiest version with smarty 3, and then proceed ahead. I would love to do the twig converstion along with some new feature as a gsoc project, for the start I will focus on converting to smarty3 so that the functions are not deprecated.

**Comment by bradymiller** [comment]
Sounds good. Agree with going to smarty3 first (even if plan to go to twig) since that will get the code in much better shape for transition to twig.

**Comment by mdsupport** [comment]
For existing installations it is going to be a surprise since there were some forms in contrib that used templates.  Does anyone have an idea if 2 and 3 can coexist in the codebase?  It would be shame to waste effort in fixing and testing those if they are not used.

**Comment by mdsupport** [comment]
@thesmallstar Just want to make sure you are aware of [the developer notes](https://raw.githubusercontent.com/smarty-php/smarty/master/SMARTY_2_BC_NOTES.txt) to be the starting point for developing your conversion strategy.

**Comment by thesmallstar** [comment]
Yes thank you so much, I bookmarked it and [this](https://www.smarty.net/v3_overview).
For the twig conversion, I will look at its feasibility and reliability and let you know accordingly.

**Comment by sunsetsystems** [comment]
The calendar bypassed Smarty in places largely because it was too slow. It really needs a complete rewrite and is probably not a good candidate for templating.

**Comment by bradymiller** [comment]
Regarding calendar, can reassess after get it working with smarty3 with the work-around (ie. allowing to keep the php tags).

**Comment by juggernautsei** [comment]
@thesmallstar Are you still working on this project?

**Comment by kojiromike** [comment]
If I'm not mistaken, the twig migration has superseded this. If I am, please feel free to reopen with comments.
